serde-support/src/main/java/io/micronaut/serde/support/serdes/EnumSerde.java

```diff
@@ -187,3 +228,106 @@ public E deserializeNullable(@NonNull Decoder decoder, @NonNull DecoderContext c
         return transform(v);
     }
 }
+
+final class EnumValueDeserializer<E extends Enum<E>> implements Deserializer<E> {
+
+    private final Argument<Object> valueType;
+    private final Deserializer<?> valueDeserializer;
+    private final EnumBeanIntrospection<? super E> enumIntrospection;
+    private final Function<E, Object> valueExtractor;
+    private final boolean allowNull;
+
+    public EnumValueDeserializer(Argument<Object> valueType, Deserializer<?> valueDeserializer, EnumBeanIntrospection<? super E> enumIntrospection, Function<E, Object> valueExtractor, boolean allowNull) {
+        this.valueType = valueType;
+        this.valueDeserializer = valueDeserializer;
+        this.enumIntrospection = enumIntrospection;
+        this.valueExtractor = valueExtractor;
+        this.allowNull = allowNull;
+    }
+
+    @NonNull
+    private E transform(@NonNull Decoder decoder, Object value) throws IOException {
+        for (EnumBeanIntrospection.EnumConstant<? super E> enumConstant : enumIntrospection.getConstants()) {
+            E enumValue = (E) enumConstant.getValue();
+            Object extractedValue = valueExtractor.apply(enumValue);
+            if (Objects.equals(extractedValue, value)) {
+                return enumValue;
+            }
+        }
+
+        var allowedValues = enumIntrospection.getConstants().stream()
+            .map(EnumBeanIntrospection.EnumConstant::getValue)
+            .map(enumValue -> valueExtractor.apply((E) enumValue))
+            .map(Object::toString)
+            .collect(Collectors.joining(", "));
+
+        throw decoder.createDeserializationException("Expected one of [%s] but was '%s'".formatted(allowedValues, value), value);
+    }
+
+    @Override
+    public E deserialize(@NonNull Decoder decoder, @NonNull DecoderContext context, @NonNull Argument<? super E> type) throws IOException {
+        return transform(decoder, valueDeserializer.deserialize(decoder, context, valueType));
+    }
+
+    @Override
+    public E deserializeNullable(@NonNull Decoder decoder, @NonNull DecoderContext context, @NonNull Argument<? super E> type) throws IOException {
+        Object v = valueDeserializer.deserializeNullable(decoder, context, valueType);
+        if (!allowNull && v == null) {
+            return null;
+        }
+        return transform(decoder, v);
+    }
+}
+
+final class EnumPropertySerializer<E extends Enum<E>> implements Serializer<E> {
+
+    private final EnumBeanIntrospection<? extends E> enumIntrospection;
+
+    EnumPropertySerializer(EnumBeanIntrospection<? extends E> enumIntrospection) {
+        this.enumIntrospection = enumIntrospection;
+    }
+
+    @Override
+    public void serialize(@NonNull Encoder encoder, @NonNull EncoderContext context, @NonNull Argument<? extends E> type, E value) throws IOException {
+        for (EnumBeanIntrospection.EnumConstant<? extends E> enumConstant : enumIntrospection.getConstants()) {
```

Comment:
Please try to precalculate constants as a map

Reply:
Added pre-calculation in 9d4e0890d6e775e04edaa008ebab5a90e8316a40